src/search/search_engines/greedy_best_first_search.cc

```diff
@@ -24,94 +24,72 @@ utils::ExitCode GreedyBestFirstSearch<PackedStateT>::search(const Task &task,
     cout << "Starting greedy best first search" << endl;
     clock_t timer_start = clock();
     SparseStatePacker state_packer(task);
-
-
-    priority_queue<Node, vector<Node>, NodeComparison> q;  // Queue has Node structures
-    segmented_vector::SegmentedVector<pair<int, LiftedOperatorId>> cheapest_parent;
-
-    segmented_vector::SegmentedVector<SparsePackedState> index_to_state;
-    unordered_map<SparsePackedState, int, PackedStateHash> visited;
-
-    segmented_vector::SegmentedVector<int> shortest_distance;
-
-    index_to_state.push_back(state_packer.pack(task.initial_state));
-    cheapest_parent.push_back(make_pair(-1, LiftedOperatorId(-1, vector<int>())));
-
-    this->heuristic_layer = heuristic.compute_heuristic(task.initial_state, task);
-    cout << "Initial heuristic value " << this->heuristic_layer << endl;
-    statistics.report_f_value_progress(this->heuristic_layer);
-
-    q.emplace(0, heuristic.compute_heuristic(task.initial_state, task), this->state_counter);
-    shortest_distance.push_back(0);
-    visited[state_packer.pack(task.initial_state)] = this->state_counter++;
-
-    if (task.is_goal(task.initial_state)) {
-        cout << "Initial state is a goal" << endl;
-        print_goal_found(generator, timer_start);
-        extract_plan(
-            cheapest_parent, state_packer.pack(task.initial_state), visited, index_to_state, state_packer, task);
-
-        return utils::ExitCode::SUCCESS;
-    }
-
-    while (not q.empty()) {
-        Node head = q.top();
-        size_t next = head.id;
-        int h = head.h;
-        int g = head.g;
-        q.pop();
-
-        statistics.report_f_value_progress(g);
+    StatePackerT packer(task);
+
+    priority_queue<GBFSNode> queue;
+
+    SearchNode& root_node = space.insert_or_get_previous_node(packer.pack(task.initial_state), LiftedOperatorId::no_operator, StateID::no_state);
+    heuristic_layer = heuristic.compute_heuristic(task.initial_state, task);
+    root_node.open(0, heuristic_layer);
+    cout << "Initial heuristic value " << heuristic_layer << endl;
+    statistics.report_f_value_progress(heuristic_layer);
+    queue.emplace(root_node.state_id, 0, heuristic_layer);
+
+    if (check_goal(task, generator, timer_start, task.initial_state, root_node, space)) return utils::ExitCode::SUCCESS;
+
+    while (not queue.empty()) {
+        const GBFSNode gbfs_n = queue.top();
+        StateID sid = gbfs_n.get_id();
+        SearchNode &node = space.get_node(sid);
+        int h = node.h;
+        int g = node.g;
+        queue.pop();
```

Comment:
Still - I'd put this line next to `queue.top()`, as it is conceptually at least one atomic operation...